### PR TITLE
feat: [DS custom pk] sync engine and async storage adapter

### PR DESCRIPTION
**Pull request by david-mcafee**
#### Description of changes

This PR updates the DataStore sync engine, async storage adapter, and async storage database to handle custom primary keys. 

There is also a minor fix for the in memory store, as `clear` was broken when not provided a callback. 

Comments have also been added to existing code to explain how things work.

#### Issue #, if available

N/A

#### Description of how you validated changes
- Manual tested with a sample app using 4 models to validate the usage of custom primary keys on all CRUD operations, including querying by PK (primary key), SK (sort key), and PK + SK.
- Updated unit tests

#### Checklist

- [ X ] PR description included
- [ X ] `yarn test` passes
- [ X ] Tests are [changed](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ X ] Relevant documentation is changed or added (and PR referenced)
    - Added in-line documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by manueliglesias — APPROVED** [review]
Let's get this merged to the epic and continue the review there

**Review by iartemiev — COMMENTED** [review]
A few minor nits, can be fixed after we merge to the epic branch

**Review by iartemiev — APPROVED** [review]

File: packages/datastore/src/storage/adapter/AsyncStorageDatabase.ts [hunk 1]
```diff
@@ -85,39 +88,54 @@ class AsyncStorageDatabase {
...
 
+		// Save item in db
 		await this.storage.setItem(itemKey, JSON.stringify(item));
 	}
 
 	async batchSave<T extends PersistentModel>(
 		storeName: string,
-		items: ModelInstanceMetadata[]
+		items: ModelInstanceMetadata[],
+		keyPath: string[]
```
[the remark this answers is not in the document]
Comment (david-mcafee) [reply packages/datastore/src/storage/adapter/AsyncStorageDatabase.ts:116]: Will do 👍

File: packages/datastore/src/storage/adapter/InMemoryStore.ts [hunk 2]
```diff
@@ -12,15 +12,17 @@ export class InMemoryStore {
 	multiRemove = async (keys: string[], callback?) => {
 		keys.forEach(k => this.db.delete(k));
 
-		callback();
+		callback ? callback() : null;
```
[the remark this answers is not in the document]
Comment (david-mcafee) [reply packages/datastore/src/storage/adapter/InMemoryStore.ts]: Done!

File: packages/datastore/src/sync/outbox.ts [hunk 3]
```diff
@@ -186,9 +196,14 @@ class MutationEventOutbox {
...
+
+		const recordId = getIdentifierValue(userModelDefinition, record);
+
 		const predicate = ModelPredicateCreator.createFromExisting<MutationEvent>(
 			mutationEventModelDefinition,
-			c => c.modelId('eq', record.id).id('ne', this.inProgressMutationEventId)
+			c => c.modelId('eq', recordId).id('ne', this.inProgressMutationEventId)
 		);
 
 		const outdatedMutations = await storage.query(
```
[the remark this answers is not in the document]
Comment (david-mcafee) [reply packages/datastore/src/sync/outbox.ts:209]: Updated

**Pull request merged**